feat: add encryption to remote_state

norman-zon opened the pull request:
add the most basic and naive implementation of an encryption attribute to the remote_state block, to generate encryption config for the backend

<!-- Prepend '[WIP]' to the title if this PR is still a work-in-progress. Remove it when it is ready for review! -->

## Description

Fixes #3495.

<!-- Description of the changes introduced by this PR. -->
This PR adds an attribute `encryption` to the `remote_state` block that generates a `encryption` block in the backend config. This allow to use the [OpenTofu state encryption](https://opentofu.org/docs/language/state/encryption/) directly from within terragrunt, thus keeping encryption config DRY.


## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [ ] Update the docs.
- [ ] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [ ] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

<!-- One-line description of the PR that can be included in the final release notes. -->
Added / Removed / Updated [X].

### Migration Guide

<!-- Important: If you made any backward incompatible changes, then you must write a migration guide! -->

yhakbar commented:
First: Great job, @norman-zon ! This is looking good!

If we're going to have another attribute on the `remote_state`, why not have it be part of the `config`? The reason I was saying that the most straightforward approach to this was to have an extra configuration block for `remote_state` was to support multiple `encryption` blocks like it is in the `terraform.backend` block.

Given all that, would you be willing to adjust the implementation so that it's either a configuration block or an attribute that's a part of `config`? If not, is there a reason that the current approach is better?

norman-zon commented:
@yhakbar thanks for the kind words.

I did not nest  `encryption` in the config block, because it seemed to me this would make it necessary to split the `config`, to clarify it creates two different `terraform` blocks . 
So this would lead to:

```hcl
remote_state {
  backend = "xxx"
  config = {
    bucket {
      name = mybucket
      prefix = ....
    }
    encryption {
       key_provider = ....
    }
}
```

Otherwise mixing up attributes for the bucket (`backend` block) and encryption (`encryption` block) in `config` seems confusing to me. But this would make it necessary to refactor the whole `config` block and be a breaking change.

Does this make sense to you?

yhakbar commented:
> @yhakbar thanks for the kind words.
> 
> I did not nest `encryption` in the config block, because it seemed to me this would make it necessary to split the `config`, to clarify it creates two different `terraform` blocks .
> ...
> Does this make sense to you?

Can you confirm that the configuration _has_ to create two `terraform` blocks? Is it the case that the following is invalid OpenTofu configuration?

```
terraform {
  backend "s3" {
    bucket         = "bucket"
    key            = "something/tf.tfstate"
  }
  encryption {
    key_provider "test" "default" {
  }
}
```

I would expect that I would prefer for them to be configured together, but I haven't used this feature, so I'm speaking out of ignorance.

norman-zon commented:
Sorry, I think I was not clear enough. We don't need to have two different `terraform` blocks, but two different blocks inside `terraform`.


> Can you confirm that the configuration _has_ to create two `terraform` blocks? Is it the case that the following is invalid OpenTofu configuration?
> 
> ```terraform
> terraform {
>   backend "s3" {
>     bucket         = "bucket"
>     key            = "something/tf.tfstate"
>   }
>   encryption {
>     key_provider "test" "default" {
>   }
> }
> ```

This is valid OpentTofu config.

For me it would be counter-intuitive, to have the `remote_state.config` block, which as of now is de-facto translated into a `backend` tofu/terraform block mixed up with attributes that have to end up in  the `encryption` tofu/terraform block. 

For example we would have something like:

```hcl
remote_state {
  backend = "xxx"
  config = {
    bucket = "my-bucket"
    key_provider = "pbkdf2"
  }
```

So one `config` block in terragrunt is the source for two different block types in tofu/terraform. I would rather separate this.

Does this clear up my rationale?

norman-zon commented:
@yhakbar , could you take a look at this again? I would love to wrap it up.

yhakbar commented:
Hey @norman-zon ,

Sorry I haven't had a ton of time to devote to reviewing this. I'll try my best to install this version locally and play with it sometime soon to be able to speak to whether it provides a good experience for a Terragrunt user.

My delay is partially due to the fact that I'm inexperienced with the OpenTofu feature.

yhakbar commented:
Hey @norman-zon ,

First, I'm taking a deeper look at this PR, and I want to say great work! This is really nice. It definitely looks like you're on the right path, and I'm gonna be trying it out locally to make sure it does work in a fixture I create.

Here are there the main issues that I'm seeing that would prevent this from being merged:

1. There is/are no added [fixture(s)](https://github.com/gruntwork-io/terragrunt/tree/main/test/fixtures) for this behavior. It really helps to have valid implementation of a new feature as a fixture to demonstrate that it works, and to explain how it can be used.
2. There is/are not added [integration test(s)](https://github.com/gruntwork-io/terragrunt/tree/main/test) for this behavior. It really helps to have integration tests for new features so that we know no regression is happening in future updates.
3. Lints are failing. It's important to ensure lints pass so that we know we're keeping code quality at a manageable level. I know you said in your issue that you're not super familiar with Golang, but [golangci-lint](https://github.com/golangci/golangci-lint) is the linter for Golang, and when adding the relevant plugin for your code editor (see [this](https://golangci-lint.run/welcome/integrations/)), you'll get live feedback as you edit to make sure you pass the lints.

    These are the failures we're seeing in CI:
    ```
    codegen/generate.go:295:3: commentFormatting: put a space between `//` and comment text (gocritic)
                    //extract key_provider first to create key_provider block
                    ^
    remote/remote_state.go:244:52: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
                    return fmt.Errorf("error creating provider: %v", err)
                                                                     ^
    remote/remote_state.go:254:57: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
                    return fmt.Errorf("error decoding struct to map: %v", err)
                                                                          ^
    remote/remote_encryption.go:28:50: the given struct should be annotated with the `mapstructure` tag (musttag)
            if err := mapstructure.Decode(encryptionConfig, &b); err != nil {
                                                            ^
    remote/remote_encryption_test.go:10:1: Function TestUnmarshalConfig missing the call to method parallel (paralleltest)
    func TestUnmarshalConfig(t *testing.T) {
    ^
    remote/remote_encryption_test.go:108:2: Range statement for test TestUnmarshalConfig missing the call to method parallel in test Run (paralleltest)
            for _, tt := range tests {
            ^
    remote/remote_encryption_test.go:124:1: Function TestToMap missing the call to method parallel (paralleltest)
    func TestToMap(t *testing.T) {
    ^
    remote/remote_encryption_test.go:185:2: Range statement for test TestToMap missing the call to method parallel in test Run (paralleltest)
            for _, tt := range tests {
            ^
    codegen/generate.go:298:16: fmt.Errorf can be replaced with errors.New (perfsprint)
                            return nil, fmt.Errorf(encryptionKeyProviderKey + " is mandatory but not found in the encryption map")
                                        ^
    remote/remote_encryption.go:81:2: ST1003: struct field KmsKeyId should be KmsKeyID (stylecheck)
            KmsKeyId                        int    `mapstructure:"kms_key_id"`
            ^
    remote/remote_encryption_test.go:201:5: require-error: for error assertions use require (testifylint)
                                    assert.NoError(t, err)
                                    ^
    codegen/generate.go:300:3: assignments should only be cuddled with other assignments (wsl)
                    keyProviderTraversal := hcl.Traversal{
                    ^
    codegen/generate.go:344:4: assignments should only be cuddled with other assignments (wsl)
                            ctyVal, err := convertValue(encryption[key])
                            ^
    codegen/generate.go:345:4: only one cuddle assignment allowed before if statement (wsl)
                            if err != nil {
                            ^
    codegen/generate.go:348:4: if statements should only be cuddled with assignments (wsl)
                            if keyProviderBlockBody != nil {
                            ^
    remote/remote_encryption.go:38:2: only one cuddle assignment allowed before if statement (wsl)
            if err != nil {
            ^
    remote/remote_encryption.go:41:2: if statements should only be cuddled with assignments (wsl)
            if err := decoder.Decode(encryptionConfig); err != nil {
            ^
    remote/remote_encryption.go:51:2: only one cuddle assignment allowed before if statement (wsl)
            if err != nil {
            ^
    remote/remote_encryption.go:54:2: return statements should not be cuddled if block has more than two lines (wsl)
            return result, nil
            ^
    config/config.go:288:3: assignments should only be cuddled with other assignments (wsl)
                    config.Encryption = remoteStateEncryption
                    ^
    ```
4. Current integration tests are failing. It can be hard to run all the tests locally, especially if you don't have an active AWS subscription, so I'll give you the names of the failing tests. You can run only the specific failing test by running `go test -run 'NAME OF TEST' ./...`. If you can't have the test pass locally because of stuff missing in the cloud or something, don't worry about it. Just know why the test is failing, and that it's not failing because of a panic or anything, and we can re-run the tests in CI to see if they pass there.

    Failing Tests:
    - `TestReadTerragruntAuthProviderCmdRemoteState`
    - `TestDependenciesOptimisation`
    - `TestRenderJSONConfig`
    - `TestDependencyMockOutputMergeWithStateDefault`
    - `TestDependencyMockOutputMergeWithStateFalse`
    - `TestOrderedMapOutputRegressions1102`
    - `TestDependencyMockOutputMergeStrategyWithStateDeepMapOnly`
    - `TestDependencyMockOutputMergeStrategyWithStateCompatFalse`
    - `TestDependencyMockOutputMergeStrategyWithStateCompatTrue`
    - `TestDependencyMockOutputMergeStrategyWithStateNoMerge`
    - `TestDependencyMockOutputMergeStrategyWithStateCompatConflict`
    - `TestDependencyMockOutputMergeStrategyWithStateDefault`
    - `TestDependencyMockOutputMergeStrategyWithStateShallow`
    - `TestTerragruntRemoteStateCodegenGeneratesBackendBlock`
    - `TestDependencyMockOutputMergeWithStateNoOverride`
    - `TestDependencyMockOutputMergeWithStateTrueNotAllowed`
    - `TestTerragruntRemoteStateCodegenDoesNotGenerateWithSkip`
    - `TestDependencyMockOutputMergeWithStateTrue`
    - `TestTerragruntRemoteStateCodegenErrorsIfExists`
    - `TestTerragruntRemoteStateCodegenOverwrites`

 
5. [Docs](https://github.com/gruntwork-io/terragrunt/tree/main/docs) aren't updated. Features that aren't documented aren't used, so we need to make sure this is well documented for folks to take advantage of it!

I know that's a lot of feedback, and I hope that's not intimidating. Tag me again if you encounter any issue you can't overcome or want another look, and I'm happy to help. Feel free to reach out in Discord too if you need a quick question answered on Golang development, etc. You might also be able to get more eyes on this from other members of the community to help you out.

norman-zon commented:
@yhakbar
Thank you for the detailed feedback, this helped a lot.
I incorporated all requested changes. 

For the integration tests to work, you would have to create a GCP and AWS KMS key and add them [here](https://github.com/gruntwork-io/terragrunt/blob/4b3d12f7405977f5f6bb467441e0fa0f686e8a92/test/integration_tofu_state_encryption_test.go#L24). 

As for how to test the encryption, checking for the `encrypted_data` values to be base64 was the best I could come up with.

The documentation is a bit scarce, I know. But it felt like everything is explained in the Tofu docs already. Again on the topic of this being a OpenTofu only feature, I was not sure how to indicate this best.

yhakbar commented:
Hey @norman-zon ,

This is looking really good! Well done, adding all of this.

Do you mind having the AWS key depend on this?
https://github.com/gruntwork-io/terragrunt/blob/4b3d12f7405977f5f6bb467441e0fa0f686e8a92/test/fixtures/sops-kms/secrets.yaml#L13

We already setup a test key to support integrations with AWS KMS. I'm syncing up with the team on the GCP side of things. We have a GCP project we put aside for Terragrunt integration tests that is used for testing remote state storage, but I don't think we leave around a test key for GCP KMS testing.

To get this moving along, I would actually ask that you just skip the GCP tests for now (`t.Skip`), with a message stating that you need to setup your own key for it to work right, so it's skipped by default. We'll make another ticket to circle back to this and add integration tests for it.

As for how to test it properly, I think you're doing the right thing. As long as you verify that the state file does indeed have base64 encoded values, and OpenTofu is continuing to work, we should be fine.

I'm gonna read through things more carefully, so you might get more feedback soon.

norman-zon commented:
Thanks again for the feedback.

I updated the PR.

Unfortunately I cannot get AWS KMS to work. I added the [key](https://github.com/gruntwork-io/terragrunt/blob/4b3d12f7405977f5f6bb467441e0fa0f686e8a92/test/fixtures/sops-kms/secrets.yaml#L13) you mentioned but get this error:

```log
=== NAME  TestTofuStateEncryptionAWSKMS
    integration_tofu_state_encryption_test.go:70: Failed to run Terragrunt command 'terragrunt apply -auto-approve --terragrunt-non-interactive --terragrunt-working-dir /var/folders/x5/ch39pzb94s503sxr04shwxbjmnk0d6/T/terragrunt-test4001692128/fixtures/tofu-state-encryption/aws-kms' due to error: util.ProcessExecutionError Failed to execute "tofu init" in /var/folders/x5/ch39pzb94s503sxr04shwxbjmnk0d6/T/terragrunt-test4001692128/fixtures/tofu-state-encryption/aws-kms
        ╷
        │ Error: Unable to build encryption key data
        │
        │ key_provider.aws_kms.default failed with error: errors were encountered in
        │ aws kms configuration
        │ No valid credential sources found : Please see
        │ https://opentofu.org/docs/language/settings/backends/s3
        │ for more information about providing credentials.
        │
        │ Error: failed to refresh cached credentials, no EC2 IMDS role found,
        │ operation error ec2imds: GetMetadata, exceeded maximum number of attempts,
        │ 3, request send failed, Get
        │ "http://169.254.169.254/latest/meta-data/iam/security-credentials/": dial
        │ tcp 169.254.169.254:80: connect: host is down
        │
```
We don't use AWS in our stack, so I have no own account to tinker around with. Maybe someone with more AWS experience can chip in here?

yhakbar commented:
Hey @norman-zon , great work! I feel like we're almost there!

I see failing unit tests here:
- `TestUnmarshalConfig/AWSKMS_invalid_config`
- `TestUnmarshalConfig/AWSKMS_full_config`
- `TestRemoteStateAsCtyDrift`

If you can fix them soon, feel free to do so, otherwise, I'll see if I can chain a PR into this before the end of the day ET to fix the tests so that we can merge your PR in. We are on company holiday starting next week, so we might not be able to merge this until next year otherwise.

norman-zon commented:
I was on holiday myself, so no worries. The last 3 tests are fixed now. Thanks for your ongoing support and patience!

yhakbar commented:
I owe you a re-review on this. Will look to get to this soon, @norman-zon !

yhakbar commented:
Seeing failures that do seem related to the changes you were trying to effect earlier:

```
    remote_encryption_test.go:224: failed to unmarshal config: failed to decode key provider: 1 error(s) decoding:
        
        * 'kms_key_id' expected type 'string', got unconvertible type 'int', value: '123456789'
=== CONT  TestDiffersFrom/gcs_null_values_ignored
[0;90m15:04:43.193[0m [0;94mDEBUG [0m Backend gcs has not changed.
=== CONT  TestToMap/PBKDF2_partial_config
=== CONT  TestToMap/GCPKMS_full_config
--- FAIL: TestToMap (0.00s)
    --- PASS: TestToMap/PBKDF2_full_config (0.00s)
    --- FAIL: TestToMap/AWSKMS_full_config (0.00s)
    --- PASS: TestToMap/PBKDF2_partial_config (0.00s)
    --- PASS: TestToMap/GCPKMS_full_config (0.00s)
```

```
=== NAME  TestReadTerragruntConfigFull
    integration_test.go:2308: 
        	Error Trace:	/home/circleci/project/test/integration_test.go:2308
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"backend":"local", "config":map[string]interface {}{"path":"foo.tfstate"}, "disable_dependency_optimization":false, "disable_init":false, "generate":map[string]interface {}{"if_exists":"overwrite_terragrunt", "path":"backend.tf"}}
        	            	actual  : map[string]interface {}{"backend":"local", "config":map[string]interface {}{"path":"foo.tfstate"}, "disable_dependency_optimization":false, "disable_init":false, "encryption":map[string]interface {}{"path":"foo.tfstate"}, "generate":map[string]interface {}{"if_exists":"overwrite_terragrunt", "path":"backend.tf"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(map[string]interface {}) (len=5) {
        	            	+(map[string]interface {}) (len=6) {
        	            	  (string) (len=7) "backend": (string) (len=5) "local",
        	            	@@ -7,2 +7,5 @@
        	            	  (string) (len=12) "disable_init": (bool) false,
        	            	+ (string) (len=10) "encryption": (map[string]interface {}) (len=1) {
        	            	+  (string) (len=4) "path": (string) (len=11) "foo.tfstate"
        	            	+ },
        	            	  (string) (len=8) "generate": (map[string]interface {}) (len=2) {
        	Test:       	TestReadTerragruntConfigFull
--- FAIL: TestReadTerragruntConfigFull (3.20s)
```

norman-zon commented:
@yhakbar thanks again for the review. I fixed the unit test, but have a hard time with `TestReadTerragruntConfigFull`. I don't understand, why the content of the `config` block gets read instead of the `encryption` block:

```log
Diff:
--- Expected
+++ Actual
@@ -8,3 +8,3 @@
(string) (len=10) "encryption": (map[string]interface {}) (len=1) {
-  (string) (len=12) "key_provider": (string) (len=3) "foo"
+  (string) (len=4) "path": (string) (len=11) "foo.tfstate"
},
```

Maybe someone can give an insight here?

yhakbar commented on config/config_as_cty.go:
```diff
@@ -622,6 +622,7 @@ func RemoteStateAsCty(remoteState *remote.RemoteState) (cty.Value, error) {
 	}
 
 	output["config"] = ctyJSONVal
+	output["encryption"] = ctyJSONVal
```
You're missing this above this line:

```go
	ctyJSONVal, err = convertToCtyWithJSON(remoteState.Encryption)
	if err != nil {
		return cty.NilVal, err
	}
```

yhakbar commented:
We've got some good failures! This is progress!

```
=== NAME  TestRenderJsonMetadataTerraform
    integration_json_test.go:475: 
        	Error Trace:	/home/circleci/project/test/integration_json_test.go:475
        	Error:      	Not equal: 
        	            	expected: "{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test216486916/fixtures/render-json-metadata/terraform-remote-state/common/remote_state.hcl\"},\"value\":{\"backend\":\"s3\",\"config\":{\"bucket\":\"mybucket\",\"key\":\"path/to/my/key\",\"region\":\"us-east-1\"},\"disable_dependency_optimization\":false,\"disable_init\":false,\"generate\":null}}"
        	            	actual  : "{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test216486916/fixtures/render-json-metadata/terraform-remote-state/common/remote_state.hcl\"},\"value\":{\"backend\":\"s3\",\"config\":{\"bucket\":\"mybucket\",\"key\":\"path/to/my/key\",\"region\":\"us-east-1\"},\"disable_dependency_optimization\":false,\"disable_init\":false,\"encryption\":null,\"generate\":null}}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-{"metadata":{"found_in_file":"/tmp/terragrunt-test216486916/fixtures/render-json-metadata/terraform-remote-state/common/remote_state.hcl"},"value":{"backend":"s3","config":{"bucket":"mybucket","key":"path/to/my/key","region":"us-east-1"},"disable_dependency_optimization":false,"disable_init":false,"generate":null}}
        	            	+{"metadata":{"found_in_file":"/tmp/terragrunt-test216486916/fixtures/render-json-metadata/terraform-remote-state/common/remote_state.hcl"},"value":{"backend":"s3","config":{"bucket":"mybucket","key":"path/to/my/key","region":"us-east-1"},"disable_dependency_optimization":false,"disable_init":false,"encryption":null,"generate":null}}
        	Test:       	TestRenderJsonMetadataTerraform
--- FAIL: TestRenderJsonMetadataTerraform (1.15s)
=== NAME  TestRenderJsonMetadataIncludes
    integration_json_test.go:328: 
        	Error Trace:	/home/circleci/project/test/integration_json_test.go:328
        	Error:      	Not equal: 
        	            	expected: "{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test1617316429/fixtures/render-json-metadata/includes/common/common.hcl\"},\"value\":{\"backend\":\"s3\",\"config\":{\"bucket\":\"mybucket\",\"key\":\"path/to/my/key\",\"region\":\"us-east-1\"},\"disable_dependency_optimization\":false,\"disable_init\":false,\"generate\":null}}"
        	            	actual  : "{\"metadata\":{\"found_in_file\":\"/tmp/terragrunt-test1617316429/fixtures/render-json-metadata/includes/common/common.hcl\"},\"value\":{\"backend\":\"s3\",\"config\":{\"bucket\":\"mybucket\",\"key\":\"path/to/my/key\",\"region\":\"us-east-1\"},\"disable_dependency_optimization\":false,\"disable_init\":false,\"encryption\":null,\"generate\":null}}"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-{"metadata":{"found_in_file":"/tmp/terragrunt-test1617316429/fixtures/render-json-metadata/includes/common/common.hcl"},"value":{"backend":"s3","config":{"bucket":"mybucket","key":"path/to/my/key","region":"us-east-1"},"disable_dependency_optimization":false,"disable_init":false,"generate":null}}
        	            	+{"metadata":{"found_in_file":"/tmp/terragrunt-test1617316429/fixtures/render-json-metadata/includes/common/common.hcl"},"value":{"backend":"s3","config":{"bucket":"mybucket","key":"path/to/my/key","region":"us-east-1"},"disable_dependency_optimization":false,"disable_init":false,"encryption":null,"generate":null}}
        	Test:       	TestRenderJsonMetadataIncludes
--- FAIL: TestRenderJsonMetadataIncludes (2.28s)
=== NAME  TestTerragruntWorksWithIncludeNoMerge
    integration_include_test.go:94: 
        	Error Trace:	/home/circleci/project/test/integration_include_test.go:274
        	            				/home/circleci/project/test/integration_include_test.go:94
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"backend":"s3", "config":map[string]interface {}{"bucket":"__INVALID_NAME__", "encrypt":true, "key":"qa/my-app/terraform.tfstate", "region":"us-west-2"}, "disable_dependency_optimization":false, "disable_init":false, "generate":interface {}(nil)}
        	            	actual  : map[string]interface {}{"backend":"s3", "config":map[string]interface {}{"bucket":"__INVALID_NAME__", "encrypt":true, "key":"qa/my-app/terraform.tfstate", "region":"us-west-2"}, "disable_dependency_optimization":false, "disable_init":false, "encryption":interface {}(nil), "generate":interface {}(nil)}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(map[string]interface {}) (len=5) {
        	            	+(map[string]interface {}) (len=6) {
        	            	  (string) (len=7) "backend": (string) (len=2) "s3",
        	            	@@ -10,2 +10,3 @@
        	            	  (string) (len=12) "disable_init": (bool) false,
        	            	+ (string) (len=10) "encryption": (interface {}) <nil>,
        	            	  (string) (len=8) "generate": (interface {}) <nil>
        	Test:       	TestTerragruntWorksWithIncludeNoMerge
--- FAIL: TestTerragruntWorksWithIncludeNoMerge (20.32s)
=== NAME  TestRenderJSONConfig
    integration_debug_test.go:211: 
        	Error Trace:	/home/circleci/project/test/integration_debug_test.go:211
        	Error:      	Not equal: 
        	            	expected: map[string]interface {}{"backend":"local", "config":map[string]interface {}{"path":"foo.tfstate"}, "disable_dependency_optimization":false, "disable_init":false, "generate":map[string]interface {}{"if_exists":"overwrite_terragrunt", "path":"backend.tf"}}
        	            	actual  : map[string]interface {}{"backend":"local", "config":map[string]interface {}{"path":"foo.tfstate"}, "disable_dependency_optimization":false, "disable_init":false, "encryption":interface {}(nil), "generate":map[string]interface {}{"if_exists":"overwrite_terragrunt", "path":"backend.tf"}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(map[string]interface {}) (len=5) {
        	            	+(map[string]interface {}) (len=6) {
        	            	  (string) (len=7) "backend": (string) (len=5) "local",
        	            	@@ -7,2 +7,3 @@
        	            	  (string) (len=12) "disable_init": (bool) false,
        	            	+ (string) (len=10) "encryption": (interface {}) <nil>,
        	            	  (string) (len=8) "generate": (map[string]interface {}) (len=2) {
        	Test:       	TestRenderJSONConfig
--- FAIL: TestRenderJSONConfig (46.99s)
```

norman-zon commented:
I just fixed the tests, by setting the assertion to "encryption": nil, . I think in the long run it would be a good idea to upgrade the fixtures to have encryption blocks to really test the desired behaviour

levkohimins commented on codegen/generate.go:
```diff
@@ -284,6 +290,80 @@ func RemoteStateConfigToTerraformCode(backend string, config map[string]interfac
+		keyProviderBlockBody := encryptionBlockBody.AppendNewBlock(encryptionKeyProviderKey, []string{keyProvider, encryptionResourceName}).Body()
+
+		// Append method block
+		methodBlock := encryptionBlockBody.AppendNewBlock("method", []string{encryptionDefaultMethod, encryptionResourceName}).Body()
```
"method" should be a constant.

levkohimins commented on codegen/generate.go:
```diff
@@ -284,6 +290,80 @@ func RemoteStateConfigToTerraformCode(backend string, config map[string]interfac
+		methodBlock.SetAttributeTraversal("keys", keyProviderTraversal)
+
+		// Append state block
+		stateBlock := encryptionBlockBody.AppendNewBlock("state", nil).Body()
```
"state" should be a constant.

levkohimins commented on codegen/generate.go:
```diff
@@ -284,6 +290,80 @@ func RemoteStateConfigToTerraformCode(backend string, config map[string]interfac
+		stateBlock.SetAttributeTraversal("method", methodTraversal)
+
+		// Append plan block
+		planBlock := encryptionBlockBody.AppendNewBlock("plan", nil).Body()
```
"plan" should be a constant.

levkohimins commented on codegen/generate.go:
```diff
@@ -284,6 +290,80 @@ func RemoteStateConfigToTerraformCode(backend string, config map[string]interfac
+
+		// Append plan block
+		planBlock := encryptionBlockBody.AppendNewBlock("plan", nil).Body()
+		planBlock.SetAttributeTraversal("method", methodTraversal)
```
"method" should be a constant.

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+type GenericRemoteEncryptionKeyProvider[T RemoteEncryptionKeyProvider] struct {
+	T T `mapstructure:",squash"`
```
```suggestion
	Data T `mapstructure:",squash"`
```
Not sure about `Data`, but definitely not `T`.

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	return nil
+}
+
+func (b *GenericRemoteEncryptionKeyProvider[T]) ToMap() (map[string]interface{}, error) {
```
```suggestion
func (b *GenericRemoteEncryptionKeyProvider[T]) ToMap() (map[string]any, error) {
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+func (b *GenericRemoteEncryptionKeyProvider[T]) ToMap() (map[string]interface{}, error) {
+	var result map[string]interface{}
```
```suggestion
	var result map[string]any
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	err := mapstructure.Decode(b.T, &result)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode struct to map: %w", err)
```
```suggestion
		return nil, errors.Errorf("failed to decode struct to map: %w", err)
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	case "aws_kms":
+		return &GenericRemoteEncryptionKeyProvider[RemoteEncryptionKeyProviderAWSKMS]{}, nil
+	default:
+		return nil, fmt.Errorf("unknown provider type: %s", providerType)
```
```suggestion
		return nil, errors.Errorf("unknown provider type: %s", providerType)
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -24,6 +24,7 @@ type RemoteState struct {
 	DisableDependencyOptimization bool                   `mapstructure:"disable_dependency_optimization" json:"DisableDependencyOptimization"`
 	Generate                      *RemoteStateGenerate   `mapstructure:"generate" json:"Generate"`
 	Config                        map[string]interface{} `mapstructure:"config" json:"Config"`
+	Encryption                    map[string]interface{} `mapstructure:"encryption" json:"Encryption"`
```
```suggestion
	Encryption                    map[string]any `mapstructure:"encryption" json:"Encryption"`
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
 	config := state.Config
 
+	// Initialize the encryption config based on the key provider
+	var encryption map[string]interface{}
```
```suggestion
	var encryption map[string]any
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
+
+	switch {
+	case state.Encryption == nil:
+		terragruntOptions.Logger.Debugf("No encryption block in remote_state config")
```
```suggestion
		terragruntOptions.Logger.Debug("No encryption block in remote_state config")
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
+	case state.Encryption == nil:
+		terragruntOptions.Logger.Debugf("No encryption block in remote_state config")
+	case len(state.Encryption) == 0:
+		terragruntOptions.Logger.Debugf("Empty encryption block in remote_state config")
```
```suggestion
		terragruntOptions.Logger.Debug("Empty encryption block in remote_state config")
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
+	case len(state.Encryption) == 0:
+		terragruntOptions.Logger.Debugf("Empty encryption block in remote_state config")
+	default:
+		keyProvider, ok := state.Encryption["key_provider"].(string)
```
We should reuse the constant.

https://github.com/norman-zon/terragrunt/blob/8c4680593667daa7f2a5d999e7a642a562574f31/codegen/generate.go#L65

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
+
+		encryptionProvider, err := NewRemoteEncryptionKeyProvider(keyProvider)
+		if err != nil {
+			return fmt.Errorf("error creating provider: %w", err)
```
```suggestion
			return errors.Errorf("error creating provider: %w", err)
```

levkohimins commented on remote/remote_state.go:
```diff
@@ -231,6 +233,36 @@ func (state *RemoteState) GenerateTerraformCode(terragruntOptions *options.Terra
+
+		encryption, err = encryptionProvider.ToMap()
+		if err != nil {
+			return fmt.Errorf("error decoding struct to map: %w", err)
```
```suggestion
			return errors.Errorf("error decoding struct to map: %w", err)
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	decoder, err := mapstructure.NewDecoder(decoderConfig)
+
+	if err != nil {
+		return fmt.Errorf("failed to create decoder: %w", err)
```
```suggestion
		return errors.Errorf("failed to create decoder: %w", err)
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	if err := decoder.Decode(encryptionConfig); err != nil {
+		return fmt.Errorf("failed to decode key provider properties: %w", err)
```
```suggestion
		return errors.Errorf("failed to decode key provider properties: %w", err)
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+func (b *GenericRemoteEncryptionKeyProvider[T]) UnmarshalConfig(encryptionConfig map[string]interface{}) error {
+	// Decode the key provider type using the default decoder config
+	if err := mapstructure.Decode(encryptionConfig, &b); err != nil {
+		return fmt.Errorf("failed to decode key provider: %w", err)
```
```suggestion
		return errors.Errorf("failed to decode key provider: %w", err)
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+	T T `mapstructure:",squash"`
+}
+
+func (b *GenericRemoteEncryptionKeyProvider[T]) UnmarshalConfig(encryptionConfig map[string]interface{}) error {
```
```suggestion
func (b *GenericRemoteEncryptionKeyProvider[T]) UnmarshalConfig(encryptionConfig map[string]any) error {
```

levkohimins commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+
+func NewRemoteEncryptionKeyProvider(providerType string) (RemoteEncryptionConfig, error) {
+	switch providerType {
+	case "pbkdf2":
```
I would recommend something like this:

```go
func (*RemoteEncryptionKeyProviderPBKDF2) Name() string {
  return "pbkdf2"
}

```

```suggestion
	case new(RemoteEncryptionKeyProviderPBKDF2).Name():
```

denis256 commented on remote/remote_encryption.go:
```diff
@@ -0,0 +1,94 @@
+		return &GenericRemoteEncryptionKeyProvider[RemoteEncryptionKeyProviderGCPKMS]{}, nil
+	case "aws_kms":
+		return &GenericRemoteEncryptionKeyProvider[RemoteEncryptionKeyProviderAWSKMS]{}, nil
+	default:
```
I think should be added test to cover this case too

denis256 commented on codegen/generate.go:
```diff
@@ -284,6 +290,80 @@ func RemoteStateConfigToTerraformCode(backend string, config map[string]interfac
 	}
 
+	// encryption can be empty
+	if len(encryption) > 0 {
```
Can changed flow to return if encryption is 0 to not have nested blocks?

yhakbar commented:
Released in [v0.72.2](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.72.2).
